openapi3filter/req_resp_decoder.go

```diff
@@ -224,20 +224,65 @@ func decodeStyledParameter(param *openapi3.Parameter, input *RequestValidationIn
 
 func decodeValue(dec valueDecoder, param string, sm *openapi3.SerializationMethod, schema *openapi3.SchemaRef) (interface{}, error) {
 	var decodeFn func(param string, sm *openapi3.SerializationMethod, schema *openapi3.SchemaRef) (interface{}, error)
-	switch schema.Value.Type {
-	case "array":
-		decodeFn = func(param string, sm *openapi3.SerializationMethod, schema *openapi3.SchemaRef) (interface{}, error) {
-			return dec.DecodeArray(param, sm, schema)
+	if len(schema.Value.AnyOf) > 0 {
+		for _, sr := range schema.Value.AnyOf {
+			value, err := decodeValue(dec, param, sm, sr)
+			if value == nil && err == nil {
+				continue
+			}
+			return value, err
+		}
+		return nil, nil
+	}
+	if len(schema.Value.AllOf) > 0 {
+		isMatched := true
+		var value interface{}
+		var err error
+		for _, sr := range schema.Value.AllOf {
+			value, err = decodeValue(dec, param, sm, sr)
+			if value == nil {
+				isMatched = false
+				break
+			}
 		}
-	case "object":
-		decodeFn = func(param string, sm *openapi3.SerializationMethod, schema *openapi3.SchemaRef) (interface{}, error) {
-			return dec.DecodeObject(param, sm, schema)
+		if isMatched == false {
+			return nil, nil
 		}
-	default:
-		decodeFn = dec.DecodePrimitive
+		return value, err
 	}
+	if len(schema.Value.OneOf) > 0 {
+		isMatched := 0
+		var value interface{}
+		var err error
+		for _, sr := range schema.Value.OneOf {
+			value, err := decodeValue(dec, param, sm, sr)
+			if value != nil || err != nil {
+				isMatched++
+			}
+		}
+		if isMatched > 1 {
```

Comment:
> oneOf – validates the value against exactly one of the subschemas

This should be `matched == 1`.

Reply:
No , i think if matched ==1 , it means this check is ok. it will return value, err  ,maybe code is 
`if isMatched == 1{
  return value, err
}
return nil,nil`

Reply:
Oh I see. Actually this `> 1` check should be returning an error, no?

Reply:
Here . i think we shouldn't return error if doesn't match one of . because if API defined optional query part within oneof . but the request URI doesn't have any query part. so in this scenario, we can't deal with it as an error.

Reply:
How does one defines an optional query part within oneOf?

Reply:
Sorry, I mean if we defined one optional(required is false) query part .and this query parameter has an object schema which included "oneof" , such i have added UT in the validation_test.go,  
`{
							Value: &openapi3.Parameter{
								In:   "query",
								Name: "queryArgOneOf",
								Schema: openapi3.NewOneOfSchema(
									openapi3.NewStringSchema().WithMaxLength(2),
									openapi3.NewDateTimeSchema(),
								).NewRef(),
							},
						},`

if i return error when oneof doesn't matech .such following codes: `if isMatched == 1 {
			return value, nil
		}
		return nil, errors.New("decode Oneof failed")`

then will appear following error , but i think we shouldn't report an error in this case. do you agree ? so i return nil instead of error.
--- FAIL: TestFilter (0.00s)
	/home/shenjie/go/src/github.com/getkin/kin-openapi/openapi3filter/validation_test.go:133: Request: POST http://example.com/api/prefix/v/suffix
	/home/shenjie/go/src/github.com/getkin/kin-openapi/openapi3filter/validation_test.go:188: 
			Error Trace:	validation_test.go:188
			Error:      	Received unexpected error:
			            	Parameter 'queryArgOneOf' in query has an error: decode Oneof failed
			Test:       	TestFilter
FAIL
FAIL	github.com/getkin/kin-openapi/openapi3filter	0.004s

Reply:
How about returning an error when `value == nil` and `param.Required == true`? I agree that if param isn't required that shouldn't be an error.

Reply:
OK，good idea ! i will update it .